FoundationDB initialization script

jeanleonov opened the pull request:
- the script downloads and installs required deb packages.
 - configures foundationdb.
 - configures systemd unit (instead of init.d).
 - runs user command on fdbcli (creation of new DB is assumed).

jeanleonov committed:
FoundationDB initialization script

 - the script downloads and installs required deb packages.
 - configures foundationdb.
 - configures systemd unit (instead of init.d).
 - runs user command on fdbcli (creation of new DB is assumed).

jeanleonov committed:
Addressing PR comments

 - Use tmpfile.d.
 - Do not accept servers number argument.
 - Do not manage memory limits.
 - Wait for FDB cluster to start.
 - Determine servers number using cluster file content.

jeanleonov committed:
Keep regular init.d supervisor

jeanleonov committed:
Use appscale cache directory for packages cache

jeanleonov committed:
download_all_artifacts.sh for bootstrap

jeanleonov committed:
README.md

jeanleonov committed:
Make sh files executable for owner

cdonati commented on foundationdb/init-foundationdb.sh:
```diff
@@ -0,0 +1,171 @@
+  fi
+done
+
+if [ -z "${HOST_SERVER_IDS}" ] ; then
```
Cluster files should only contain a list of coordination servers, so the absence of the current machine's IP from a cluster file should not indicate that the current machine should not be part of the cluster.

The typical procedure for adding additional machines to an existing cluster is taking a cluster file from a machine in the current cluster and starting fdb on another machine with that cluster file. I think it's a bit unusual for a cluster file to be defined before the cluster exists since the idea is that they are meant to be descriptive of a given cluster rather than prescriptive about how a cluster should be.

A multinode setup seems like it would be a job for an external system coordinating the work.

cdonati commented on foundationdb/init-foundationdb.sh:
```diff
@@ -0,0 +1,171 @@
+fi
+export HOST_TO_LISTEN_ON
+export DATA_DIR
+export FDB_SERVERS=$(
```
A single storage server/coordinator seems fine. As it's currently written this looks like it's trying to start a storage server for every coordinator listed in the cluster file. I don't think it ever makes sense to have more than one coordinator on the same machine.

cdonati commented on foundationdb/init-foundationdb.sh:
```diff
@@ -0,0 +1,171 @@
+
+log 'Waiting for FDB cluster to start'
+wait_start=$(date +%s)
+while ! fdbcli --exec status
```
I think this will return a successful exit code even if the database is unavailable.

cdonati reviewed (commented):
Rather than having one script for downloading and other for installing and running, I think it makes more sense to have one script for installing (downloading if necessary) and the other for running.

Things can go wrong during the installation process, so I think it would be preferable to avoid doing so during runtime. I would also be fine with having the start script fail when the packages haven't been installed in order to reduce the work of the start script.

jeanleonov committed:
Addressing PR comments

 - Download and install in one script.
 - Configure and start in another script.

jeanleonov committed:
Merge branch 'master' into fdb-initialization

jeanleonov committed:
Enforce public_address in .conf file

jeanleonov committed:
Merge branch 'fdb-initialization' of github.com:jeanleonov/appscale-thirdparties into fdb-initialization

jeanleonov committed:
Test if fdb is running using systemctl

sjones4 commented on common.sh:
```diff
@@ -0,0 +1,46 @@
+    expected_md5="$2"
+    mkdir -p ${PACKAGE_CACHE}
+    if [ -f ${cached_file} ]; then
+        md5=($(md5sum ${cached_file}))
```
We may want to move to `sha384sum`.

sjones4 commented on install_all.sh:
```diff
@@ -1,3 +1,9 @@
 
+SCRIPT_DIR="$( realpath --strip "$( dirname "${BASH_SOURCE[0]}" )" )"
+
+for script in $(find "${SCRIPT_DIR}" -name install.sh) ; do
```
Could check only subdirectories e.g.`for script in "${SCRIPT_DIR}"/*/install.sh`

sjones4 commented on foundationdb/install.sh:
```diff
@@ -0,0 +1,26 @@
+
+log "Installing ${FDB_SERVER_PKG}"
+dpkg --install ${PACKAGE_CACHE}/foundationdb-server_6.1.8-1_amd64.deb
+
```
Should the cached packages be deleted after install?

sjones4 commented on foundationdb/configure-and-start-fdb.sh:
```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env bash
+# init-foundationdb.sh script ensures that FoundationDB server
```
> This script ensures that FoundationDB server

sjones4 commented on foundationdb/configure-and-start-fdb.sh:
```diff
@@ -0,0 +1,114 @@
+
+log 'Waiting for FDB service to start'
+wait_start=$(date +%s)
+while ! systemctl is-active --quiet foundationdb.service
```
I would expect this to always be true, since we successfully started the service.

sjones4 reviewed (approved):
You could use policy (`/usr/sbin/policy-rc.d`) to prevent installed packages from running services.

jeanleonov committed:
Addressing PR comments

 - using sha384 instead of md5.
 - updated comments in scripts.
 - not checking if fdb service is active.
 - remove cached package.

jeanleonov commented:
Thanks for comments @sjones4. I addressed all but `/usr/sbin/policy-rc.d`, - I didn't understand how to use it in an explicit way. Instructions which I could find showed that I need to modify `policy-rc.d` executable before installing package. I couldn't find how can I explicitly specify which service shouldn't be started.

cdonati commented:
Unfortunately, I don't think the package we use for the server install respects policy-rc.d. Future packages likely will, though, now that https://github.com/apple/foundationdb/pull/2001 is merged.
